src/inline_subgraph.cpp

```diff
@@ -0,0 +1,160 @@
+                if(last->name() == "@return")
+                {
+                    mod_outputs = last->inputs();
+                }
+                else
+                {
+                    // no return instruction, last instruction is output
+                    mod_outputs.push_back(last);
+                }
+
+                // only one output is considered for now
+                for(const auto& out : mod_outputs)
+                {
+                    auto mod_out = out->eval();
+                    if(mod_out.empty())
+                        return;
+                    arg_outs.push_back(mod_out);
+                }
+            }
+
+            auto out_num            = arg_outs.size() / 2;
+            auto cond               = ins->inputs().front();
+            const auto& ins_outputs = ins->outputs();
+            auto icond              = p.insert_instruction(
+                ins, make_op("convert", {{"target_type", shape::float_type}}), cond);
+            for(std::size_t i = 0; i < out_num; ++i)
+            {
+                auto l0 = p.add_literal(literal(arg_outs.at(i).get_shape(), arg_outs.at(i).data()));
+                auto l1 = p.add_literal(
+                    literal(arg_outs.at(i + out_num).get_shape(), arg_outs.at(i + out_num).data()));
+                auto lens  = l0->get_shape().lens();
+                auto mcond = p.insert_instruction(
+                    ins, make_op("multibroadcast", {{"output_lens", lens}}), icond);
+                std::size_t elem_num = l0->get_shape().elements();
+                std::vector<float> vec_offset(elem_num, elem_num);
+                std::vector<float> vec_ind(elem_num);
+                shape sidx{shape::float_type, lens};
+                std::iota(vec_ind.begin(), vec_ind.end(), elem_num);
+                auto lidx    = p.add_literal(literal(sidx, vec_ind));
+                auto loffset = p.add_literal(literal(sidx, vec_offset));
+                auto moffset = p.insert_instruction(ins, make_op("mul"), mcond, loffset);
+                auto f_ind   = p.insert_instruction(ins, make_op("sub"), lidx, moffset);
+                auto ins_ind = p.insert_instruction(
+                    ins, make_op("convert", {{"target_type", shape::int32_type}}), f_ind);
+                auto l01 = p.insert_instruction(ins, make_op("concat", {{"axis", 0}}), l0, l1);
+                auto rl  = p.insert_instruction(
+                    ins, make_op("reshape", {{"dims", {l0->get_shape().elements() * 2}}}), l01);
+                auto r = p.insert_instruction(ins, make_op("gather", {{"axis", 0}}), rl, ins_ind);
+                p.replace_instruction(ins_outputs.at(i), r);
+            }
+        }
+        // cond is constant, inline the corresponding subgraph and discard the other one
+        else
+        {
+            inline_submodule(p, ins);
+        }
+    }
+}
+
+void inline_subgraph::inline_submodule(module& p, instruction_ref ins) const
```

Comment:
This can be a free function defined only in the cpp file.

Reply:
Also should be `module& m`.

Reply:
fixed